
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>MToolBox-variant-calling &#8212; MToolBox  documentation</title>
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Run MToolBox" href="run-the-pipeline.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="run-the-pipeline.html" title="Run MToolBox"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">MToolBox  documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="run-the-pipeline.html" accesskey="U">Run MToolBox</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">MToolBox-variant-calling</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="mtoolbox-variant-calling">
<h1>MToolBox-variant-calling<a class="headerlink" href="#mtoolbox-variant-calling" title="Permalink to this headline">¶</a></h1>
<p>This wrapper performs QC, quality trimming of raw reads, read alignment, alignment filtering, variant calling. The final output is a VCF file.</p>
<div class="section" id="what-should-i-look-to-here">
<h2>What should I look to, here?<a class="headerlink" href="#what-should-i-look-to-here" title="Permalink to this headline">¶</a></h2>
<p><strong>TL;DR</strong></p>
<ul class="simple">
<li><p>the VCF file in the <code class="code docutils literal notranslate"><span class="pre">results/vcf</span></code> folder</p></li>
<li><p>the BED file(s) in the <code class="code docutils literal notranslate"><span class="pre">results/&lt;sample&gt;</span></code> folder.</p></li>
</ul>
<p>Both file formats can be imported in a genome browser (<em>eg</em> IGV) to visually inspect your results.</p>
<div class="section" id="the-vcf-output">
<h3>The VCF output<a class="headerlink" href="#the-vcf-output" title="Permalink to this headline">¶</a></h3>
<p>The VCF (variant call format) file is roughly a table where, after a ton of comment lines (starting with <code class="code docutils literal notranslate"><span class="pre">##</span></code>), rows are variants and columns are samples. You can find a detailed description of the VCF format <a class="reference external" href="https://www.internationalgenome.org/wiki/Analysis/vcf4.0">here</a>, although MToolBox-snakemake provides a slightly different version to report the allele heteroplasmy frequency. To spare you some headache, we’ll give you a brief summary of the genotype info you’ll find for each sample.</p>
<p>The <code class="code docutils literal notranslate"><span class="pre">FORMAT</span></code> field lists data types and order that are available for samples (following fields). Each sample has colon-separated data corresponding to the types specified in the <code class="code docutils literal notranslate"><span class="pre">FORMAT</span></code>.</p>
<ul class="simple">
<li><p><code class="code docutils literal notranslate"><span class="pre">GT</span></code> (genotype) reports all the alleles found for that sample, where <code class="code docutils literal notranslate"><span class="pre">0</span></code> is the reference allele (<code class="code docutils literal notranslate"><span class="pre">REF</span></code> field) and <code class="code docutils literal notranslate"><span class="pre">1</span></code>, <code class="code docutils literal notranslate"><span class="pre">2</span></code>, … are the alleles in the same order as in the <code class="code docutils literal notranslate"><span class="pre">ALT</span></code> field.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">DP</span></code> (depth) reports the total coverage depth for that site in the genome, <em>i.e.</em> the number of reads mapping on it.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">HF</span></code> (heteroplasmic frequency) reports, for each allele in <code class="code docutils literal notranslate"><span class="pre">GT</span></code> (excluding <code class="code docutils literal notranslate"><span class="pre">REF</span></code>), the HF.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">CILOW</span></code> (confidence interval, lower bound) reports, for each allele in <code class="code docutils literal notranslate"><span class="pre">GT</span></code> (excluding <code class="code docutils literal notranslate"><span class="pre">REF</span></code>), the lower bound of the CI.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">CIUP</span></code> (confidence interval, upper bound) reports, for each allele in <code class="code docutils literal notranslate"><span class="pre">GT</span></code> (excluding <code class="code docutils literal notranslate"><span class="pre">REF</span></code>), the upper bound of the CI.</p></li>
<li><p><code class="code docutils literal notranslate"><span class="pre">SDP</span></code> (strand read depth) reports, for each allele in <code class="code docutils literal notranslate"><span class="pre">GT</span></code> (excluding <code class="code docutils literal notranslate"><span class="pre">REF</span></code>), the number of times the allele was observed on the plus strand and on the minus strand, semi-colon separated.</p></li>
</ul>
<p>If you consider this example:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1">#CHROM       POS    ID  REF  ALT  QUAL  FILTER  INFO       FORMAT                   Scer_mt_500K                     Scer_mt_100K</span>
NC_001224.1  <span class="m">50000</span>  .   A    C    .     PASS    <span class="nv">AN</span><span class="o">=</span><span class="m">4</span><span class="p">;</span><span class="nv">AC</span><span class="o">=</span><span class="m">2</span>  GT:DP:HF:CILOW:CIUP:SDP  <span class="m">0</span>/1:359:0.46:0.409:0.511:90<span class="p">;</span><span class="m">75</span>   <span class="m">0</span>/1:75:0.387:0.284:0.5:11<span class="p">;</span><span class="m">18</span>
NC_001224.1  <span class="m">69045</span>  .   C    T    .     PASS    <span class="nv">AN</span><span class="o">=</span><span class="m">2</span><span class="p">;</span><span class="nv">AC</span><span class="o">=</span><span class="m">1</span>  GT:DP:HF:CILOW:CIUP:SDP  <span class="m">0</span>/1:365:0.033:0.018:0.057:0<span class="p">;</span><span class="m">12</span>   ./.:.:.:.:.:.
</pre></div>
</div>
<p>sample Scer_mt_500K, in mt position 50000, has 359 aligned reads and one variant allele (<code class="code docutils literal notranslate"><span class="pre">C</span></code>) with HF=0.46. The CI for this HF is 0.409 to 0.511. The variant allele is supported by 90 reads on the plus strand and 75 reads on the minus strand.</p>
</div>
<div class="section" id="the-bed-output">
<h3>The BED output<a class="headerlink" href="#the-bed-output" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference external" href="https://m.ensembl.org/info/website/upload/bed.html">BED (browser extensible data) file</a> is a useful and intuitive way to inspect the variant calling results through a genome browser. Once you import this file in a genome browser, variants will be colour-coded (blue for mutations, green for insertions, red for deletions) and shaded according to the HF (the darker the shade, the higher the HF).</p>
</div>
<div class="section" id="what-to-do-next">
<h3>What to do next?<a class="headerlink" href="#what-to-do-next" title="Permalink to this headline">¶</a></h3>
<p>Once you have run the wrapper, you will notice that the <code class="code docutils literal notranslate"><span class="pre">results</span></code> folder includes a ton of files. A guide to these files is coming soon.</p>
</div>
</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">MToolBox-variant-calling</a><ul>
<li><a class="reference internal" href="#what-should-i-look-to-here">What should I look to, here?</a><ul>
<li><a class="reference internal" href="#the-vcf-output">The VCF output</a></li>
<li><a class="reference internal" href="#the-bed-output">The BED output</a></li>
<li><a class="reference internal" href="#what-to-do-next">What to do next?</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="run-the-pipeline.html"
                        title="previous chapter">Run MToolBox</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/mtoolbox-variant-calling.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="run-the-pipeline.html" title="Run MToolBox"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">MToolBox  documentation</a> &#187;</li>
          <li class="nav-item nav-item-1"><a href="run-the-pipeline.html" >Run MToolBox</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">MToolBox-variant-calling</a></li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2019, Domenico Simone.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 3.2.1.
    </div>
  </body>
</html>